File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/OfflineRouteRetrievalTask.java
package com.mapbox.services.android.navigation.v5.navigation;

import android.os.AsyncTask;
import android.support.annotation.NonNull;

import com.google.gson.Gson;
import com.mapbox.api.directions.v5.models.DirectionsRoute;
import com.mapbox.navigator.Navigator;
import com.mapbox.navigator.RouterResult;

import timber.log.Timber;

class OfflineRouteRetrievalTask extends AsyncTask<OfflineRoute, Void, DirectionsRoute> {
  private static final int FIRST_ROUTE = 0;
  private final Navigator navigator;
  private final OnOfflineRouteFoundCallback callback;
  private RouterResult routerResult;

  OfflineRouteRetrievalTask(Navigator navigator, OnOfflineRouteFoundCallback callback) {
    this.navigator = navigator;
    this.callback = callback;
  }

  // For testing only
  OfflineRouteRetrievalTask(Navigator navigator, OnOfflineRouteFoundCallback callback, RouterResult routerResult) {
    this(navigator, callback);
    this.routerResult = routerResult;
  }

  @Override
  protected DirectionsRoute doInBackground(OfflineRoute... offlineRoutes) {
    String url = offlineRoutes[FIRST_ROUTE].buildUrl();

    synchronized (navigator) {
      routerResult = navigator.getRoute(url);
    }

    return offlineRoutes[FIRST_ROUTE].retrieveOfflineRoute(routerResult);
  }

  @Override
  protected void onPostExecute(DirectionsRoute offlineRoute) {
    if (offlineRoute != null) {
      callback.onRouteFound(offlineRoute);
    } else {
      String errorMessage = generateErrorMessage();
      OfflineError error = new OfflineError(errorMessage);
      callback.onError(error);
    }
  }

  @NonNull
  private String generateErrorMessage() {
    String jsonResponse = routerResult.getJson();
    Gson gson = new Gson();
    OfflineRouteError routeError = gson.fromJson(jsonResponse, OfflineRouteError.class);
    String errorMessage = String.format("Error occurred fetching offline route: %s - Code: %d", routeError.getError(),
      routeError.getErrorCode());
    Timber.e(errorMessage);
    return errorMessage;
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/OfflineRouteRetrievalTask.kt
package com.mapbox.services.android.navigation.v5.navigation

import android.os.AsyncTask
import com.google.gson.Gson
import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.navigator.Navigator
import com.mapbox.navigator.RouterResult
import timber.log.Timber

internal class OfflineRouteRetrievalTask(
    private val navigator: Navigator,
    private val callback: OnOfflineRouteFoundCallback
) : AsyncTask<OfflineRoute, Void, DirectionsRoute>() {

    @Volatile private lateinit var routerResult: RouterResult

    // For testing only
    internal constructor(
        navigator: Navigator,
        callback: OnOfflineRouteFoundCallback,
        routerResult: RouterResult
    ) : this(navigator, callback) {
        this.routerResult = routerResult
    }

    companion object {
        private const val FIRST_ROUTE = 0
    }

    override fun doInBackground(vararg offlineRoutes: OfflineRoute): DirectionsRoute? {
        val url = offlineRoutes[FIRST_ROUTE].buildUrl()

        synchronized(navigator) {
            navigator.getRoute(url) {}
        }

        return offlineRoutes[FIRST_ROUTE].retrieveOfflineRoute(routerResult)
    }

    public override fun onPostExecute(offlineRoute: DirectionsRoute?) {
        if (offlineRoute != null) {
            callback.onRouteFound(offlineRoute)
        } else {
            callback.onError(OfflineError(generateErrorMessage()))
        }
    }

    private fun generateErrorMessage(): String {
        val (_, _, error, errorCode) = Gson().fromJson(routerResult.json, OfflineRouteError::class.java)

        val errorMessage = "Error occurred fetching offline route: $error - Code: $errorCode"

        Timber.e(errorMessage)
        return errorMessage
    }
}
